### PR TITLE
feat(inputs.prometheus): Add HTTP service discovery support

### PR DESCRIPTION
**Pull request by djbarajas**
## Summary
This is a basic implementation to support HTTP service discovery. There are many use cases where something like this could be useful such as when hosts change and a developer might need to write their own http service discovery service. I tested this with a basic implementation of http service discovery I wrote here https://github.com/djbarajas/http-prometheus-sd/tree/main.

## Checklist

- [x] No AI generated code was used in this PR

## Related issues

resolves https://github.com/influxdata/telegraf/issues/16695.

### CODE REVIEW

File: plugins/inputs/prometheus/README.md [hunk 1]
```diff
@@ -186,6 +186,12 @@ and `bearer_token_string` option. See the
   #     [inputs.prometheus.consul.query.tags]
   #       host = "{{.Node}}"
 
+  ## Scrape Services available with http service discovery
+  # [inputs.prometheus.http_service_discovery]
+  #   enabled = true
+  #   http_sd_url = "http://localhost:9000/service-discovery"
```
Comment (djbarajas) [inline plugins/inputs/prometheus/README.md:192]: I am not supporting auth for the service discovery endpoint because I do not necessarily need this right now

File: plugins/inputs/prometheus/README.md [hunk 2]
```diff
@@ -186,6 +186,12 @@ and `bearer_token_string` option. See the
   #     [inputs.prometheus.consul.query.tags]
   #       host = "{{.Node}}"
```
Comment (danielbarajas-rescale) [inline plugins/inputs/prometheus/README.md:187]: The reason this is failing the lint readme is because it has been missing metadata. Let me know if that must be fixed to get this change merged.

Comment (djbarajas) [inline plugins/inputs/prometheus/README.md:187]: Looks like this input plugin has been missing some metadata which is why its not passing the lint check. Let me know if that is required to fix to get this pr merged

**Review by Hipska — APPROVED** [review]

File: plugins/inputs/prometheus/sample.conf [hunk 3]
```diff
@@ -161,6 +161,12 @@
   #     [inputs.prometheus.consul.query.tags]
   #       host = "{{.Node}}"
 
+  ## Scrape Hosts available with http service discovery
+  # [inputs.prometheus.http_service_discovery]
+  #   enabled = true
+  #   http_sd_url = "http://localhost:9000/service-discovery"
```
[the remark this answers is not in the document]
Comment (djbarajas) [reply plugins/inputs/prometheus/sample.conf]: that makes sense to me thank you!

File: plugins/inputs/prometheus/http_service_discovery.go [hunk 4]
```diff
@@ -0,0 +1,137 @@
...
+			return err
+		}
+	} else {
+		body, err = io.ReadAll(resp.Body)
+		if err != nil {
+			return err
+		}
+	}
+
+	var targets []httpSDOutput
```
Comment (djbarajas) [inline plugins/inputs/prometheus/http_service_discovery.go:107]: there was a slight change here. I had the output wrong a bit with the standard. This should fix it

File: plugins/inputs/prometheus/README.md (hunk 2 above)
Comment (Hipska) [reply plugins/inputs/prometheus/README.md:187]: Yea, I think they will want that added. Would be nice if you add it.

File: plugins/inputs/prometheus/sample.conf [hunk 5]
```diff
@@ -161,6 +161,12 @@
   #     [inputs.prometheus.consul.query.tags]
   #       host = "{{.Node}}"
 
+  ## Scrape Hosts available with http service discovery
+  # [inputs.prometheus.http_service_discovery]
+  #   enabled = true
+  #   url = "http://localhost:9000/service-discovery"
+  #   query_interval = "5m"
```
Comment (Hipska) [inline plugins/inputs/prometheus/sample.conf:168]: There is the rule that commented out values should be representing the default values. This does not seem to be the case here. (I agree it seems also not to be the case for the consul example above)

File: plugins/inputs/prometheus/http_service_discovery.go (hunk 4 above)
Comment (Hipska) [reply plugins/inputs/prometheus/http_service_discovery.go:107]: I would have named the var a bit different to not cause confusion afterwards..

**Review by Hipska — CHANGES_REQUESTED** [review]

File: plugins/inputs/prometheus/README.md (hunk 2 above)
Comment (djbarajas) [reply plugins/inputs/prometheus/README.md:192]: After some digging I found the version it was introduced. Added it https://github.com/influxdata/telegraf/commit/5390a8ea71ca3055507cd78bdf1b1db895626f37

File: plugins/inputs/prometheus/sample.conf (hunk 5 above)
Comment (djbarajas) [reply plugins/inputs/prometheus/sample.conf]: I think this makes sense to default all values except `enabled` should be false. I have updated this. Let me know if this is getting confusing though.

**Review by Hipska — APPROVED** [review]

File: plugins/inputs/prometheus/http_service_discovery_test.go [hunk 6]
```diff
@@ -0,0 +1,61 @@
...
+	"github.com/stretchr/testify/require"
+
+	"github.com/influxdata/telegraf/testutil"
+)
+
+type FakeHTTPClient struct {
+	testing *testing.T
+}
+
+func (client *FakeHTTPClient) Do(req *http.Request) (*http.Response, error) {
```
[the remark this answers is not in the document]
Comment (srebhan) [reply plugins/inputs/prometheus/http_service_discovery_test.go:20]: Please do not fake the client but rather fake the server! This way, we do not need to introduce test-specific code and indirections into the actual implementation! The [`httptest` server](https://pkg.go.dev/net/http/httptest) is really easy to use... Also I suggest to implement "test-case" based tests here where you place the information in a subdirectory like e.g. in [this plugin](https://github.com/influxdata/telegraf/blob/9465e7182a3e5783b3ec1b777fd95cf9b69f1c65/plugins/inputs/neoom_beaam/neoom_beaam_test.go#L21).

File: plugins/inputs/prometheus/prometheus.go [hunk 7]
```diff
@@ -83,6 +83,9 @@ type Prometheus struct {
 	// Consul discovery
 	ConsulConfig consulConfig `toml:"consul"`
 
+	// http service discovery
```
Comment (srebhan) [inline plugins/inputs/prometheus/prometheus.go:86]:
```suggestion
	// HTTP service discovery
```

**Review by srebhan — COMMENTED** [review]
Thank you for your contribution @djbarajas! I do have some comments in the code...

File: plugins/inputs/prometheus/http_service_discovery.go [hunk 8]
```diff
@@ -0,0 +1,149 @@
...
+
+	var sdOutput []httpSDOutput
+	if err := json.Unmarshal(body, &sdOutput); err != nil { // Parse []byte to go struct pointer
+		return err
+	}
+
+	for _, sdOutputItem := range sdOutput {
+		for _, targetValue := range sdOutputItem.Targets {
+			// http service discovery returns <host>:<port> pairs so we default to appending http to all hosts
+			targetValue = "http://" + targetValue
```
[the remark this answers is not in the document]
Comment (djbarajas) [reply plugins/inputs/prometheus/http_service_discovery.go:127]: There might be cases where https is required but not for me. It would be simple to support both via configuration values if needed

File: plugins/inputs/prometheus/http_service_discovery.go [hunk 9]
```diff
@@ -0,0 +1,149 @@
...
+
+	var body []byte
+	if p.ContentLengthLimit != 0 {
+		limit := int64(p.ContentLengthLimit)
+		lr := io.LimitReader(resp.Body, limit)
+		body, err = io.ReadAll(lr)
+		if err != nil {
+			return err
+		}
+	} else {
```
[the remark this answers is not in the document]
Comment (djbarajas) [reply plugins/inputs/prometheus/http_service_discovery.go:108]: sounds good to me i will remove

File: plugins/inputs/prometheus/http_service_discovery.go [hunk 10]
```diff
@@ -0,0 +1,149 @@
...
+		}
+	}()
+
+	return nil
+}
+
+func (p *Prometheus) refreshHTTPServices(sdURL *url.URL, client HTTPClient) error {
+	refreshHTTPServices := make(map[string]urlAndAddress)
+
+	p.Log.Debugf("Refreshing HTTP services")
```
[the remark this answers is not in the document]
Comment (djbarajas) [reply plugins/inputs/prometheus/http_service_discovery.go]: done

File: plugins/inputs/prometheus/http_service_discovery.go [hunk 11]
```diff
@@ -0,0 +1,149 @@
...
+	}
+
+	if p.HTTPSDConfig.URL == "" {
+		p.HTTPSDConfig.URL = "http://localhost:9000/service-discovery"
+	}
+
+	httpSDUrl, err := url.Parse(p.HTTPSDConfig.URL)
+	if err != nil {
+		return fmt.Errorf("cannot parse the http service discovery url: %w", err)
+	}
```
[the remark this answers is not in the document]
Comment (djbarajas) [reply plugins/inputs/prometheus/http_service_discovery.go:47]: thats a good point. I was mostly just verifying it was a valid url before the goroutine however there are other ways I could do this without forcing the parse at the beginning and maybe its ok to just not check beforehand and let it fail

File: plugins/inputs/prometheus/http_service_discovery_test.go [hunk 12]
```diff
@@ -0,0 +1,53 @@
...
+	"os"
+	"path/filepath"
+	"testing"
+
+	"github.com/stretchr/testify/require"
+
+	"github.com/influxdata/telegraf/config"
+)
+
+func TestHttpSD(t *testing.T) {
```
Comment (djbarajas) [inline plugins/inputs/prometheus/http_service_discovery_test.go:15]: Im testing the happy path here, I could add a couple more tests if needed such as empty and invalid result from service discovery. Let me know if that is required

File: plugins/inputs/prometheus/http_service_discovery.go [hunk 13]
```diff
@@ -0,0 +1,127 @@
...
+				url:         targetURL,
+				originalURL: targetURL,
+				// in this case target labels should just be added to the tags
+				tags: sdOutputItem.Labels,
+			}
+			services[service.url.String()] = service
+		}
+	}
+
+	p.lock.Lock()
```
Comment (djbarajas) [inline plugins/inputs/prometheus/http_service_discovery.go:122]: While I locked and unlocked like the consol example, let me know if we prefer I defer unlock like in this other example https://github.com/influxdata/telegraf/blob/4c23243ed0b204a2636d5aabe3d3a6c3f89a98a1/testutil/accumulator.go#L62

File: plugins/inputs/prometheus/testcases/service_discovery/http-services.json [hunk 14]
```diff
@@ -0,0 +1,20 @@
+[
+  {
+    "targets": [
+      "localhost:9090"
+    ],
+    "labels": {
+      "host": "localhost",
+      "port": "9090"
+    }
```
Comment (Hipska) [inline plugins/inputs/prometheus/testcases/service_discovery/http-services.json:9]: Labels here look totally different than the examples from the docs: https://prometheus.io/docs/prometheus/latest/http_sd/#http_sd-format

File: plugins/inputs/prometheus/http_service_discovery.go [hunk 15]
```diff
@@ -0,0 +1,127 @@
...
+	defer resp.Body.Close()
+
+	if resp.StatusCode != http.StatusOK {
+		return fmt.Errorf("discovery failed with status %q", resp.Status)
+	}
+
+	var body []byte
+	body, err = io.ReadAll(resp.Body)
+	if err != nil {
+		return err
```
Comment (srebhan) [inline plugins/inputs/prometheus/http_service_discovery.go:94]:
```suggestion
		return fmt.Errorf("reading response body failed: %w", err)
```

File: plugins/inputs/prometheus/http_service_discovery.go (hunk 13 above)
Comment (srebhan) [reply plugins/inputs/prometheus/http_service_discovery.go:122]: It's okay as is because there is no error path and the code during which the lock is held is simple. Otherwise, for more complex code we prefer a deferred call to unlock...

File: plugins/inputs/prometheus/http_service_discovery_test.go [hunk 16]
```diff
@@ -0,0 +1,53 @@
...
+
+	// refresh http services
+	client := &http.Client{}
+	defer client.CloseIdleConnections()
+	require.NoError(t, plugin.refreshHTTPServices(server.URL, client))
+
+	plugin.lock.Lock()
+	// check we have 2 http services
+	require.Len(t, plugin.httpServices, 2)
+	plugin.lock.Unlock()
```
Comment (srebhan) [inline plugins/inputs/prometheus/http_service_discovery_test.go:52]: Here you need to defer the unlock as the `require` call can exit the function with the lock held...

File: plugins/inputs/prometheus/testcases/service_discovery/http-services.json (hunk 14 above)
Comment (srebhan) [reply plugins/inputs/prometheus/testcases/service_discovery/http-services.json:9]: I do agree with @Hipska! Ideally you should use the example from the prometheus  docs here...

Comment (djbarajas) [reply plugins/inputs/prometheus/testcases/service_discovery/http-services.json:9]: No objections to that. Makes sense to me.

Comment (djbarajas) [reply plugins/inputs/prometheus/testcases/service_discovery/http-services.json]: I don't think so. For my case we want the labels to show up in metrics as is and the docs do not mention removing that prefix.

File: plugins/inputs/prometheus/http_service_discovery.go (hunk 8 above)
[the remark this answers is not in the document]
Comment (srebhan) [reply plugins/inputs/prometheus/http_service_discovery.go:105]: Hmmm how would that be possible? Say you have a set of targets with `http` and another set with `https`... How do you know what to do here?

**Review by srebhan — APPROVED** [review]
Looks good to me. Thanks for your contribution @djbarajas!

File: plugins/inputs/prometheus/http_service_discovery.go (hunk 8 above)
[the remark this answers is not in the document]
Comment (djbarajas) [reply plugins/inputs/prometheus/http_service_discovery.go:105]: You could be correct here. The docs aren't the most clear on this piece. I think it is ok to deal with this later. Might require testing both since the docs do not have the protocol in the result unfortunately even though it is needed to make the request

File: plugins/inputs/prometheus/http_service_discovery_test.go [hunk 17]
```diff
@@ -0,0 +1,53 @@
...
+	require.NoError(t, err)
+
+	// Create a fake API server
+	server := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, _ *http.Request) {
+		if _, err := w.Write(result); err != nil {
+			w.WriteHeader(http.StatusInternalServerError)
+			t.Error(err)
+			return
+		}
+	}))
```
Comment (skartikey) [inline plugins/inputs/prometheus/http_service_discovery_test.go:31]: If w.Write() fails, calling w.WriteHeader() after the write attempt is ineffective because the HTTP headers would have already been sent.

```suggestion
	server := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
    // Verify it's a GET request
    if r.Method != http.MethodGET {
        w.WriteHeader(http.StatusMethodNotAllowed)
        return
    }
    
    w.Header().Set("Content-Type", "application/json")
    if _, err := w.Write(result); err != nil {
        t.Errorf("Failed to write response: %v", err)
        return
    }
}))
```

File: plugins/inputs/prometheus/http_service_discovery.go (hunk 8 above)
[the remark this answers is not in the document]
Comment (djbarajas) [reply plugins/inputs/prometheus/http_service_discovery.go:105]: That sounds good. I was thinking about that because it was the safe option, however I only decided against that after seeing all the examples not have the prefix. Will add that back

File: plugins/inputs/prometheus/http_service_discovery.go [hunk 18]
```diff
@@ -0,0 +1,127 @@
...
+	var body []byte
+	body, err = io.ReadAll(resp.Body)
+	if err != nil {
+		return fmt.Errorf("reading response body failed: %w", err)
+	}
+
+	var result []httpSDOutput
+	if err := json.Unmarshal(body, &result); err != nil {
+		return fmt.Errorf("unmarshalling JSON failed: %w", err)
+	}
```
[the remark this answers is not in the document]
Comment (djbarajas) [reply plugins/inputs/prometheus/http_service_discovery.go:101]: sure that makes sense.

**Review by skartikey — APPROVED** [review]
@djbarajas Thanks for the contribution!

**Comment by kartik162000** [comment]
@djbarajas  when this feature we are releasing

**Comment by kartik162000** [comment]
@Hipska just want to know do we have anything related to autodiscovery other than this method in docker swarm environment. My use case is scrapping the application metrics from container having replicas, so auto discovery can help there, I have gone through the documentation but haven't found anything other than this, wanted to re-ensure on that
